### PR TITLE
fix: filter metrics from hidden dimension

### PR DESCRIPTION
**Pull request by rephus**
Closes: #8765

### Description:

Before: I filter a dimension, this dimension has a derived metric, this metric is not filtered

Screenshot from 2024-03-05 09-21-14

Screenshot from 2024-03-05 09-21-42

AFter: 

Metrics and dimensions are both filtered
Screenshot from 2024-03-05 09-47-16

Once I enable user attribute, both appear again

Screenshot from 2024-03-05 09-47-40

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/services/UserAttributesService/UserAttributeUtils.ts [hunk 1]
```diff
@@ -82,14 +96,24 @@ export const getFilteredExplore = (
...
-                                    (tableReference) =>
-                                        filteredTableNames.includes(
-                                            tableReference,
-                                        ),
-                                ),
+                            ([metricName, metric]) => {
+                                const parentDimensionIsHidden =
+                                    hiddenDimensions.some(
+                                        (dimension) =>
+                                            dimension.sql === metric.sql,
```
Comment (rephus) [inline packages/backend/src/services/UserAttributesService/UserAttributeUtils.ts:103]: afaik this is the most reliable  way to know if a metric comes from a dimension, right ? 
Screenshot from 2024-03-05 09-42-25

File: examples/full-jaffle-shop-demo/dbt/models/customers.sql [hunk 2]
```diff
@@ -59,6 +59,7 @@ final as (
         customers.customer_id,
         customers.first_name,
         customers.last_name,
+        30 as age, -- fixed age is filtered using required_attributes on schema.yml
```
Comment (rephus) [inline examples/full-jaffle-shop-demo/dbt/models/customers.sql:62]: added for testing, but we can keep it there, I've added a new "fixed" field to not break anything else, let me know if you want me to remove it or  replace it

File: packages/backend/src/services/UserAttributesService/UserAttributeUtils.ts (hunk 1 above)
Comment (ZeRego) [reply packages/backend/src/services/UserAttributesService/UserAttributeUtils.ts:103]: Yes, but won't work for metrics with custom SQL. To support those, the metrics should have a `dimensionReferences: Array<{table: string; name: string;}>;` similar to the `tablesTeferences`.

Comment (rephus) [reply packages/backend/src/services/UserAttributesService/UserAttributeUtils.ts:103]: Yeah, I've been trying to figure out a way to link dimensions in metrics, but at least on the exploreCompiler, metrics and dimension are not linked, we might need to look before that happens, perhaps at `convertExplores` on `translator` ... I'll keep looking :eyes:

File: packages/backend/src/services/UserAttributesService/UserAttributeUtils.ts [hunk 3]
```diff
@@ -82,14 +96,25 @@ export const getFilteredExplore = (
...
-                                    (tableReference) =>
-                                        filteredTableNames.includes(
-                                            tableReference,
-                                        ),
-                                ),
+                            ([metricName, metric]) => {
+                                const parentDimensionIsHidden =
+                                    hiddenDimensions.some(
+                                        (dimension) =>
+                                            dimension.name ===
```
Comment (rephus) [inline packages/backend/src/services/UserAttributesService/UserAttributeUtils.ts:103]: not it checks against the metric.dimensionReference

Comment (rephus) [reply packages/backend/src/services/UserAttributesService/UserAttributeUtils.ts:103]: Screenshot from 2024-03-06 10-34-30

Comment (ZeRego) [reply packages/backend/src/services/UserAttributesService/UserAttributeUtils.ts:103]: This is still going to cause us problems as I can reference the dimension A in metrics defined under dimension B.

I propose update`exploreCompiled.ts` (pseudo code):

compileDimensionReference() - now returns the referenced dimension
```
        return {
            ...
            dimension: referencedDimension,
        };
```

compileMetricSql() - aggregates all the dimension referenced and then returns 
```
let dimensionsReferences =  new Set<{id: string, table: string, name: string}>();

...

                if(compiledReference.dimension) {
                    dimensionsReferences = ([
                        ...dimensionsReferences,
                        { id: getItemId(compiledReference.dimension), table: compiledReference.dimension.table, name: compiledReference.dimension.name}
                    ]);
                }
                
...
 
                 return { sql: compiledSql, tablesReferences, dimensionsReferences };
```

Using `{id: string, table: string, name: string}` so we have all the information possible to relate fields/tables.

**Review by ZeRego — APPROVED** [review]

**Pull request merged**